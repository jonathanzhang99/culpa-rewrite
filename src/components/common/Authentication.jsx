import PropTypes from "prop-types";
import React, { useContext, useReducer } from "react";
import { Route, Redirect } from "react-router-dom";

export const AuthContext = React.createContext(null);

function authReducer(state, action) {
  switch (action.type) {
    case "LOGIN_START":
      return {
        ...state,
        isLoading: true,
      };
    case "LOGIN_ERROR":
      return {
        ...state,
        isError: true,
        isLoading: false,
      };

    case "LOGIN_SUCCESS":
      localStorage.setItem("authToken", JSON.stringify(action.payload));
      return {
        ...state,
        isError: false,
        isLoading: false,
        isLoggedIn: true,
        authToken: action.payload,
      };

    case "LOGIN_FAILURE":
      return {
        ...state,
        isError: false,
        isLoading: false,
        isLoggedIn: false,
      };
    case "LOGOUT_SUCCESS":
      localStorage.removeItem("authToken");
      return {
        ...state,
        isError: false,
        isLoading: false,
        isLoggedIn: false,
        authToken: undefined,
      };
    default:
      throw new Error();
  }
}

const propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element),
  ]).isRequired,
};

/**
 * AuthProvider createa React Context with values such that children elements
 * in the tree can access user auth token and manipulate them
 * (e.g. login, logout)
 *
 */
export function AuthProvider({ children }) {
  const existingToken = JSON.parse(localStorage.getItem("authToken"));

  const [state, dispatch] = useReducer(authReducer, {
    isError: false,
    isLoading: false,
    isLoggedIn: !!existingToken,
    authToken: existingToken,
  });

  // `login` either returns an error or null if a token was succesfully
  // retrieved. The error will always be a string message
  const login = async ({ username, password }) => {
    dispatch({ type: "LOGIN_START" });
    const response = await fetch("/api/auth/login", {
      method: "POST",
      body: JSON.stringify({
        username,
        password,
      }),
      headers: {
        "Content-Type": "application/json",
      },
    });
    try {
      const result = await response.json();

      // Require that the backend sends either a token or an error
      result.error
        ? dispatch({ type: "LOGIN_FAILURE" })
        : dispatch({ type: "LOGIN_SUCCESS", payload: result.token });
      return result;
    } catch (err) {
      dispatch({ type: "LOGIN_ERROR" });
      return err;
    }
  };

  // `logout` simply removes the token from the client. It does NOT
  // expire the token.
  const logout = async () => {
    if (state.authToken) {
      dispatch({ type: "LOGOUT_SUCCESS" });
    }
  };

  return (
    <AuthContext.Provider
      value={{
        ...state,
        login,
        logout,
      }}
    >
      {children}
    </AuthContext.Provider>
  );
}

AuthProvider.propTypes = propTypes;

/**
 * ProtectedRoute allows access to the url only if a user is logged in.
 *
 */
export function ProtectedRoute(props) {
  const { isLoggedIn } = useContext(AuthContext);
  // eslint-disable-next-line react/jsx-props-no-spreading
  return isLoggedIn ? <Route {...props} /> : <Redirect to="/" />;
}
