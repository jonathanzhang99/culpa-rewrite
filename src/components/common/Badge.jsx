import PropTypes from "prop-types";
import React from "react";
import { Image } from "semantic-ui-react";

import bronzeNugget from "icons/bronze-nugget.svg";
import goldNugget from "icons/gold-nugget.svg";
import silverNugget from "icons/silver-nugget.svg";

const propTypesBadge = {
  badgeId: PropTypes.number.isRequired,
};

export default function Badge({ badgeId }) {
  switch (badgeId) {
    case 1: {
      return <Image avatar spaced="left" src={goldNugget} />;
    }
    case 2: {
      return <Image avatar spaced="left" src={silverNugget} />;
    }
    case 3: {
      return <Image avatar spaced="left" src={bronzeNugget} />;
    }
    default: {
      return null;
    }
  }
}

Badge.propTypes = propTypesBadge;
