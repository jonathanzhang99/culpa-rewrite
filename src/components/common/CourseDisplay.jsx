import PropTypes from "prop-types";
import React from "react";
import { Link } from "react-router-dom";
import { Header } from "semantic-ui-react";

const propTypesCourseDisplayName = {
  as: PropTypes.string,
  name: PropTypes.string.isRequired,
  code: PropTypes.string,
};

const propTypesCourseDisplayLink = {
  as: PropTypes.string,
  courseId: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  code: PropTypes.string,
};

const defaultProps = {
  as: "p",
  code: "",
};

export function CourseDisplayName({ as, code, name }) {
  return (
    <Header as={as}>
      {code} {name}
    </Header>
  );
}

export function CourseDisplayLink({ as, courseId, code, name }) {
  return (
    <Link
      to={{
        pathname: `/courses/${courseId}`,
      }}
    >
      <CourseDisplayName as={as} code={code} name={name} />
    </Link>
  );
}

CourseDisplayName.propTypes = propTypesCourseDisplayName;
CourseDisplayName.defaultProps = defaultProps;

CourseDisplayLink.propTypes = propTypesCourseDisplayLink;
CourseDisplayLink.defaultProps = defaultProps;
