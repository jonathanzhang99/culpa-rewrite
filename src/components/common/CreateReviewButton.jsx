/* eslint-disable react/jsx-props-no-spreading */
import PropTypes from "prop-types";
import React from "react";
import { Link } from "react-router-dom";
import { Button } from "semantic-ui-react";

const propTypes = {
  courseId: PropTypes.number,
  professorId: PropTypes.number,
};

const defaultProps = {
  courseId: undefined,
  professorId: undefined,
};

export default function CreateReviewButton({ courseId, professorId, ...rest }) {
  return (
    <Link
      to={{
        pathname: "/review/submit",
        state: { courseId, professorId },
      }}
    >
      <Button {...rest} />
    </Link>
  );
}

CreateReviewButton.propTypes = propTypes;
CreateReviewButton.defaultProps = defaultProps;
