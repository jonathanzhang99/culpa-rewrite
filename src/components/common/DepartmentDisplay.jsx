import PropTypes from "prop-types";
import React from "react";
import { Link } from "react-router-dom";
import { Header } from "semantic-ui-react";

const propTypesDepartmentDisplayName = {
  as: PropTypes.oneOf(["header", "text"]),
  departmentName: PropTypes.string.isRequired,
};

const propTypesDepartmentDisplayLink = {
  as: PropTypes.oneOf(["header", "text"]),
  departmentId: PropTypes.number.isRequired,
  departmentName: PropTypes.string.isRequired,
};

const defaultProps = {
  as: "text",
};

export function DepartmentDisplayName({ as, departmentName }) {
  const DepartmentName = as === "header" ? Header : "span";
  return <DepartmentName> {departmentName} </DepartmentName>;
}

export function DepartmentDisplayLink({ as, departmentId, departmentName }) {
  return (
    <Link to={`/department/${departmentId}`}>
      <DepartmentDisplayName as={as} departmentName={departmentName} />
    </Link>
  );
}

DepartmentDisplayName.propTypes = propTypesDepartmentDisplayName;
DepartmentDisplayName.defaultProps = defaultProps;

DepartmentDisplayLink.propTypes = propTypesDepartmentDisplayLink;
DepartmentDisplayLink.defaultProps = defaultProps;
