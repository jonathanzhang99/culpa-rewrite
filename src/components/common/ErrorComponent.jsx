import React from "react";

export default function ErrorComponent() {
  return <div>error occurred</div>;
}
