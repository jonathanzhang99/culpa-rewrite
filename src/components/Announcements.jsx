import PropTypes from "prop-types";
import React from "react";

import ErrorComponent from "components/common/ErrorComponent";
import LoadingComponent from "components/common/LoadingComponent";
import useDataFetch from "components/common/useDataFetch";

const propTypes = {
  messages: PropTypes.arrayOf(PropTypes.string).isRequired,
};

export function AnnouncementsSection(props) {
  // TODO: Implement Announcements
  const { messages } = props;
  return <div>{messages}</div>;
}

AnnouncementsSection.propTypes = propTypes;

export default function Announcements() {
  const {
    data: { messages },
    isLoading,
    isError,
  } = useDataFetch("/api/announcements/all", {
    messages: [],
  });

  if (isLoading || isError) {
    return isLoading ? <LoadingComponent /> : <ErrorComponent />;
  }

  return <AnnouncementsSection messages={messages} />;
}
