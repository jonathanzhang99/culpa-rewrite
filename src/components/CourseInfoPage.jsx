import PropTypes from "prop-types";
import React, { useState } from "react";
import { useParams } from "react-router-dom";
import { Icon, Accordion, Table } from "semantic-ui-react";

import { CourseDisplayName } from "components/common/CourseDisplay";
import CreateReviewButton from "components/common/CreateReviewButton";
import { DepartmentDisplayLink } from "components/common/DepartmentDisplay";
import ErrorComponent from "components/common/ErrorComponent";
import LoadingComponent from "components/common/LoadingComponent";
import { ProfessorDisplayLink } from "components/common/ProfessorDisplay";
import useDataFetch from "components/common/useDataFetch";

const propTypes = {
  courseId: PropTypes.number.isRequired,
  courseName: PropTypes.string.isRequired,
  courseCallNumber: PropTypes.string.isRequired,
  departmentId: PropTypes.number.isRequired,
  departmentName: PropTypes.string.isRequired,
  courseProfessors: PropTypes.arrayOf(
    PropTypes.shape({
      professorDepartmentId: PropTypes.number,
      professorDepartmentName: PropTypes.string,
    })
  ).isRequired,
};

export function CourseSummary({
  courseId,
  courseName,
  courseCallNumber,
  departmentId,
  departmentName,
  courseProfessors,
}) {
  const [isAccordionActive, setAccordionActive] = useState(false);
  return (
    <>
      <CourseDisplayName
        as="header"
        courseCallNumber={courseCallNumber}
        courseName={courseName}
      />
      Department:
      <DepartmentDisplayLink
        departmentId={departmentId}
        departmentName={departmentName}
      />
      <Accordion>
        <Accordion.Title
          active={isAccordionActive}
          onClick={() => setAccordionActive(!isAccordionActive)}
        >
          <Icon name={!isAccordionActive ? "angle down" : "angle up"} />
          {!isAccordionActive ? "Show" : "Hide"} all professors who teach this
          course
        </Accordion.Title>
        <Accordion.Content active={isAccordionActive}>
          <Table>
            <tbody>
              {/* Having <tbody> prevents browser bugs from inserting into the Table. */}
              {courseProfessors.map(
                ({
                  firstName,
                  lastName,
                  professorId,
                  professorDepartments,
                }) => {
                  return (
                    <Table.Row key={professorId}>
                      <Table.Cell key={`${professorId}_link`}>
                        <ProfessorDisplayLink
                          firstName={firstName}
                          lastName={lastName}
                          professorId={professorId}
                        />
                      </Table.Cell>
                      <Table.Cell key={`${professorId}_departments`}>
                        {professorDepartments.map(
                          (
                            { professorDepartmentId, professorDepartmentName },
                            index
                          ) => {
                            return (
                              <span key={professorDepartmentId}>
                                <DepartmentDisplayLink
                                  departmentId={professorDepartmentId}
                                  departmentName={professorDepartmentName}
                                />
                                {professorDepartments.length - 1 !== index
                                  ? ", "
                                  : ""}
                              </span>
                            );
                          }
                        )}
                      </Table.Cell>
                    </Table.Row>
                  );
                }
              )}
            </tbody>
          </Table>
        </Accordion.Content>
      </Accordion>
      <CreateReviewButton color="yellow" courseId={courseId.toString()}>
        WRITE A REVIEW FOR {courseName}
      </CreateReviewButton>
    </>
  );
}

export default function CourseInfoPage() {
  const { courseId } = useParams();
  const {
    data: {
      courseName,
      courseCallNumber,
      departmentId,
      departmentName,
      courseProfessors,
    },
    isLoading,
    isError,
  } = useDataFetch(`/api/course/${courseId}`, {
    courseName: "",
    courseCallNumber: "",
    departmentId: 0,
    departmentName: "",
    courseProfessors: [],
  });
  // TODO: load and return Review Summary data here

  if (isLoading || isError) {
    return isLoading ? <LoadingComponent /> : <ErrorComponent />;
  }

  return (
    <CourseSummary
      courseCallNumber={courseCallNumber}
      courseId={courseId}
      courseName={courseName}
      courseProfessors={courseProfessors}
      departmentId={departmentId}
      departmentName={departmentName}
    />
  );
}

CourseSummary.propTypes = propTypes;
