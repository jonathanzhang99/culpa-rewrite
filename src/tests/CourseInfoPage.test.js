import { render, fireEvent, screen } from "@testing-library/react";
import React from "react";
import { MemoryRouter, Route } from "react-router-dom";

import { AuthContext } from "components/common/Authentication";
import { CourseSummary } from "components/CourseInfoPage";

describe("CourseInfoPage Component", () => {
  const testCases = [
    {
      testName: "renders course info with no professors",
      courseId: 0,
      courseName: "Machine Learning",
      courseCallNumber: "COMS 4771",
      departmentId: 1,
      departmentName: "Computer Science",
      courseProfessors: [],
    },
    {
      testName: "renders course info with professor",
      courseId: 1,
      courseName: "Machine Learning",
      courseCallNumber: "COMS 4771",
      departmentId: 1,
      departmentName: "Computer Science",
      courseProfessors: [
        {
          firstName: "Nakul",
          lastName: "Verma",
          professorId: 1,
          professorDepartments: [
            {
              professorDepartmentId: 1,
              professorDepartmentName: "Computer Science",
            },
            {
              professorDepartmentId: 3,
              professorDepartmentName: "Mathematics",
            },
          ],
        },
      ],
    },
  ];

  testCases.forEach(
    ({
      testName,
      courseId,
      courseName,
      courseCallNumber,
      departmentId,
      departmentName,
      courseProfessors,
    }) => {
      test(testName, () => {
        const snapshot = render(
          <MemoryRouter>
            <CourseSummary
              courseCallNumber={courseCallNumber}
              courseId={courseId}
              courseName={courseName}
              courseProfessors={courseProfessors}
              departmentId={departmentId}
              departmentName={departmentName}
            />
          </MemoryRouter>
        );
        screen.debug();

        expect(snapshot).toMatchSnapshot();
      });
    }
  );

  /* Semantic-UI React does not change HTML for accordion folding/unfolding so 
   we test accordion functionality by checking for change in button string. */
  test("renders course info with working professor accordion", () => {
    const courseId = 1;
    const courseName = "Machine Learning";
    const courseCallNumber = "COMS 4771";
    const departmentId = 1;
    const departmentName = "Computer Science";
    const courseProfessors = [
      {
        firstName: "Nakul",
        lastName: "Verma",
        professorId: 1,
        professorDepartments: [
          {
            professorDepartmentId: 1,
            professorDepartmentName: "Computer Science",
          },
          {
            professorDepartmentId: 3,
            professorDepartmentName: "Mathematics",
          },
        ],
      },
    ];

    render(
      <MemoryRouter>
        <CourseSummary
          courseCallNumber={courseCallNumber}
          courseId={courseId}
          courseName={courseName}
          courseProfessors={courseProfessors}
          departmentId={departmentId}
          departmentName={departmentName}
        />
      </MemoryRouter>
    );

    screen.debug();
    expect(screen.queryByText(/show all professors/i)).toBeInTheDocument();

    fireEvent.click(screen.getByText(/show all professors/i));
    expect(screen.queryByText(/hide all professors/i)).toBeInTheDocument();

    fireEvent.click(screen.getByText(/hide all professors/i));
    expect(screen.queryByText(/show all professors/i)).toBeInTheDocument();
  });
});
