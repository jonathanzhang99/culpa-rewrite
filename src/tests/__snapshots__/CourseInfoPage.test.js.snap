// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseInfoPage Component renders course info with no professors 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ui header"
      >
        COMS 4771
         
        Machine Learning
      </div>
      Department:
      <a
        href="/department/1"
      >
        <span>
           
          Computer Science
           
        </span>
      </a>
      <div
        class="accordion ui"
      >
        <div
          class="title"
        >
          <i
            aria-hidden="true"
            class="angle down icon"
          />
          Show
           all professors who teach this course
        </div>
        <div
          class="content"
        >
          <table
            class="ui table"
          >
            <tbody />
          </table>
        </div>
      </div>
      <a
        href="/review"
      >
        <button
          class="ui yellow button"
        >
          WRITE A REVIEW FOR 
          Machine Learning
        </button>
      </a>
    </div>
  </body>,
  "container": <div>
    <div
      class="ui header"
    >
      COMS 4771
       
      Machine Learning
    </div>
    Department:
    <a
      href="/department/1"
    >
      <span>
         
        Computer Science
         
      </span>
    </a>
    <div
      class="accordion ui"
    >
      <div
        class="title"
      >
        <i
          aria-hidden="true"
          class="angle down icon"
        />
        Show
         all professors who teach this course
      </div>
      <div
        class="content"
      >
        <table
          class="ui table"
        >
          <tbody />
        </table>
      </div>
    </div>
    <a
      href="/review"
    >
      <button
        class="ui yellow button"
      >
        WRITE A REVIEW FOR 
        Machine Learning
      </button>
    </a>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`CourseInfoPage Component renders course info with professor 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ui header"
      >
        COMS 4771
         
        Machine Learning
      </div>
      Department:
      <a
        href="/department/1"
      >
        <span>
           
          Computer Science
           
        </span>
      </a>
      <div
        class="accordion ui"
      >
        <div
          class="title"
        >
          <i
            aria-hidden="true"
            class="angle down icon"
          />
          Show
           all professors who teach this course
        </div>
        <div
          class="content"
        >
          <table
            class="ui table"
          >
            <tbody>
              <tr
                class=""
              >
                <td
                  class=""
                >
                  <a
                    href="/professor/1"
                  >
                    <span>
                      Nakul
                       
                      Verma
                    </span>
                  </a>
                </td>
                <td
                  class=""
                >
                  <span>
                    <a
                      href="/department/1"
                    >
                      <span>
                         
                        Computer Science
                         
                      </span>
                    </a>
                    , 
                  </span>
                  <span>
                    <a
                      href="/department/3"
                    >
                      <span>
                         
                        Mathematics
                         
                      </span>
                    </a>
                    
                  </span>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
      <a
        href="/review"
      >
        <button
          class="ui yellow button"
        >
          WRITE A REVIEW FOR 
          Machine Learning
        </button>
      </a>
    </div>
  </body>,
  "container": <div>
    <div
      class="ui header"
    >
      COMS 4771
       
      Machine Learning
    </div>
    Department:
    <a
      href="/department/1"
    >
      <span>
         
        Computer Science
         
      </span>
    </a>
    <div
      class="accordion ui"
    >
      <div
        class="title"
      >
        <i
          aria-hidden="true"
          class="angle down icon"
        />
        Show
         all professors who teach this course
      </div>
      <div
        class="content"
      >
        <table
          class="ui table"
        >
          <tbody>
            <tr
              class=""
            >
              <td
                class=""
              >
                <a
                  href="/professor/1"
                >
                  <span>
                    Nakul
                     
                    Verma
                  </span>
                </a>
              </td>
              <td
                class=""
              >
                <span>
                  <a
                    href="/department/1"
                  >
                    <span>
                       
                      Computer Science
                       
                    </span>
                  </a>
                  , 
                </span>
                <span>
                  <a
                    href="/department/3"
                  >
                    <span>
                       
                      Mathematics
                       
                    </span>
                  </a>
                  
                </span>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <a
      href="/review"
    >
      <button
        class="ui yellow button"
      >
        WRITE A REVIEW FOR 
        Machine Learning
      </button>
    </a>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
